Firestore/core/test/firebase/firestore/remote/serializer_test.cc

```diff
@@ -258,7 +277,113 @@ TEST_F(SerializerTest, WritesNestedObjects) {
   ExpectRoundTrip(model, proto, FieldValue::Type::Object);
 }
 
+TEST_F(SerializerTest, BadNullValue) {
+  std::vector<uint8_t> bytes{
+      0x58,  // encoded null tag
```

Comment:
I couldn't come up with a real good way to specify the bytes in a readable manner, so I ended up just specifying them as raw numbers with comments. This is roughly calculated via: `(field_number) << 3 | wiretype`. The result is then fed to protobuf's encode varint function which will leave it alone if its <= 0x7f; otherwise, it'll encode it further, resulting in multiple bytes. (See https://developers.google.com/protocol-buffers/docs/encoding#varints for further details.)

Anyways, I think it's ok as is, even if there are a lot of magic bytes, but feel free to suggest further improvements.

Reply:
A way I've seen work before is to serialize down to bytes in the usual way and then perform targeted mutations to ensure coverage over the cases we care about, e.g. by incrementing/decrementing a valid value. In this case the serializer would generate the "encoded null tag" for us and would cut down on some of the repetition of legal values that allow us to exercise non-first-byte errors.

Separately a fuzz test can help verify that we don't crash: set up a loop for some number of iterations, generate a random number of random bytes and accept either success or an error. Feel free to tackle fuzzing separately.

Reply:
I considered that, but didn't like it as much as the targeted mutations were kinda magic. (As opposed to the raw bytes being magic. shrug.) But let me try it out and see what it looks like.

Re fuzzer. Good idea. :)

Reply:
Re. fuzzing - I think it would be awesome if we could add [libfuzzer](http://llvm.org/docs/LibFuzzer.html) to Firestore at some later point.

Reply:
Ok; check out `BadStringValue` (original style) and `BadStringValue2` (new style). WDYT?